Initial documentation for doc comment references (#6080)
dart-lang/site-www: src/content/tools/doc-comment-references.md

---
title: Documentation comment references
short-title: Comment references
description: Learn about doc comment references and their syntax.
---

Doc comments can contain references to various identifiers. Elements, such as
functions and classes, can be referenced by wrapping their name in square
brackets (`[...]`) in a doc comment (a comment starting with `///`). Examples:

```dart
/// Returns a [String].
String f() => 'Hello';

/// Wraps [object] with [Future.value].
Future<T> g<T>(T object) => Future.value(object);
```

These doc comments contain references to the `String` class, the `object`
parameter, and the `Future.value` constructor.

## Features of references

There are several benefits to referring to code elements with doc comment
references:

> ### Editor features

Author: Love it

Doc comment references enable several IDE features:

- **Code completion**
  An element's name can be code-completed within square brackets.
- **Rename refactoring**
  When an element is renamed via an IDE command, it can rewrite uses of that

>   element, including references in doc comments.

Author: Done

- **Find references**
  When an IDE lists all "references" to an element, it can include references in
  doc comments.
- **Go to definition**
  An IDE can also provide Go-to-definition support at the location of a doc
  comment reference.

:::tip

> The [comment_references](https://dart.dev/lints/comment_references) lint rule

Author: Done

can help to ensure that doc comment references are valid, avoiding typos and
mis-uses. Keeping doc comment references valid ensures that these IDE features
are enabled for each reference.
:::

### API documentation

In API documentation generated by [`dart doc`](/tools/dart-doc), a doc comment reference is
linked, if possible, to the documentation page of the element being referenced.
If the element does not have a documentation page (for example, a function

> parameter, a type variable, or a private class), then no link is created.

Reviewer: ```suggestion
parameter, a type parameter, or a private class), then no link is created.
```

Author: Done

    </td>
    <td>
Reference with their name, if they are in-scope, prefixed with an import prefix if appropriate.

```dart
/// [List] is in scope.
/// So is [math.max].
```
    </td>
  </tr>
</table>

Most library members can be referenced in a doc comment, including classes,
constants, enums, named extensions, extension types, functions, mixins, and
type aliases. This includes all in-scope library members, either declared
locally, or imported. Library members that are imported with an import prefix
can be referenced with the prefix. For example:

```dart
import 'dart:math' as math;

/// [List] is in scope.
/// So is [math.max].
int x = 7;
```

Most members of a class, an enum, an extension, an extension type, and a mixin
can also be referenced. A reference to a member that is not in scope must be
qualified (prefixed) with its container's name. For example the `wait` static
method on the `Future` class can be referenced in a doc comment with
`[Future.wait]`. This is true for instance members as well; the `add` method
and the `length` property on the `List` class can be referenced with
`[List.add]` and `[List.length]`. In-scope members like these can even be

> referenced without the qualifying container name:

Reviewer: "In-scope members like these" makes it sound like the qualified examples in the previous sentence are "in-scope". Perhaps clarify with a conditional clause, something like:

```
When container members are in-scope, such as in an instance method's doc comment, they can be referenced without the qualifying container name:
```

Author: I like it. Done.

```dart
abstract class MyList<E> implements List<E> {
  /// Refer to [add] and [contains], which is declared on [Iterable].
  void myMethod() {}
}
```

Unnamed constructors can be referenced by using the `new` name, similar to the
tear-off of an unnamed constructor. For example, `[DateTime.new]` is a
reference to the unnamed `DateTime` constructor.

Parameters of a function and parameters of a function type can be referenced in
a doc comment only when they are in scope. They can therefore only be
referenced within a doc comment on such a parameter's function or on a type
alias for such a parameter's enclosing function type.

Type parameters can be referenced in a doc comment only when they are in scope.
Therefore a type parameter of a method, top-level function, or type alias can
only be referenced within a doc comment on that element, and a type parameter
of a class, enum, extension, extension type, and mixin can only be referenced
within a doc comment on that element or on one of its members.

The doc comment for a type alias that aliases a class, enum, extension type,

> or mixin can't reference any of the aliased type's members, if it has any, as
> if they were in scope.

Reviewer: I feel the aside of "if it has any" separates the "as if they were in scope" too much. I think it's fine to just drop it:

```suggestion
or mixin can't reference any of the aliased type's members as if they were in scope.
```

Author: Yeah good call. Done.

---
title: Documentation comment references
short-title: Comment references
description: Learn about doc comment references and their syntax.
---

Doc comments can contain references to various code elements. A library member
or class member, etc. can be referenced by wrapping its name in square brackets

> (`[...]`) in a doc comment (a comment starting with `///`). Examples:

Reviewer 2: Consider adding " or `/**`", given that it's technically correct.

```dart
/// Returns a [String].
String f() => 'Hello';

/// Wraps [object] with [Future.value].
Future<T> g<T>(T object) => Future.value(object);
```

These doc comments contain references to the `String` class, the `object`
parameter, and the `Future.value` constructor.

## Features of references

There are several benefits to referring to a code element in a doc comment as a
doc comment reference:

### Editor features

Doc comment references enable several IDE features. When an element is renamed
via an IDE command, it can rewrite uses of that element, including references
in doc comments. When an IDE lists all "references" to an element, it can
include references in doc comments. An IDE can also provide Go-to-definition

> support at the location of a doc comment reference.

Reviewer 2: It mIght be worth mentioning code completion within the reference.

Author: Ah ha! Great catch.

constants, enums, named extensions, extension types, functions, mixins, and
type aliases. This includes all in-scope library members, either declared
locally, or imported. Library members which are imported with an import prefix
can be referenced with the prefix. Examples:

```dart
import 'dart:math' as math;

/// [List] is in scope.
/// So is [math.max].
int x = 7;
```

Most members of a class, an enum, an extension, an extension type, and a mixin
can also be referenced. A reference to a member that is not in scope must be
qualified with its container's name. For example the `wait` static method on
the `Future` class can be referenced in a doc comment with `[Future.wait]`.
This is true for instance members as well; the `add` method and the `length`
property on the `List` class can be referenced with `[List.add]` and
`[List.length]`. In-scope members like these can even be referenced without the
qualifying container name:

```dart
abstract class MyList<E> implements List<E> {
  /// Refer to [add] and [contains], which is declared on [Iterable].
  void myMethod() {}
}
```

The doc comment for a function can include references to the function's

> parameters.

Reviewer 2: Presumably also the type parameters of the function?

Author: Yes, that is in the next sentence.

These are a little awkward, I want to say "The type parameters of things-that-have-type-parameters can be referenced" and "The parameters of things-that-have-parameters can be referenced." But it's not as general as referencing, for example, a static method on a class: you can refer to `Future.then` anywhere, but the _only_ place from which you can refer to a parameter or a type parameter when it is in scope; You cannot refer to the `Future.value. onValue` if it is out-of-scope.

OK, I've re-worded these to focus on "parameter" and "type parameter" and expanded with a few sentences.

---
title: Documentation comment references
short-title: Comment references
description: Learn about doc comment references and their syntax.
---

> Doc comments can contain references to various code elements. A library member

Reviewer: Describing doc comment references as referring to "elements" makes sense to me and I believe is what we've always said, but is that because of its implementation ties to the analyzer and our familiarity with it? I'm not necessarily saying we shouldn't use the term, but I just want to make sure it's the best term to use.

I can find very few cases in user-facing docs where we use the term "element" like this. Never do we say "code element". In fact, we are careful to use the term as it already has a few meanings (elements in a collection, collection literal elements, HTML elements, etc).

As a side note, [`comment_references`](https://dart.dev/lints/comment_references) lint docs mention "in-scope identifiers".

@bwilkerson for thoughts based on writing diagnostic messages.

\cc @leafpetersen For other considerations :)

Author: I'll stick to "identifiers" instead of "code elements" for now. Good call. I think "element" might be more correct in other cases. For example, an _identifier_ is not referred to, a _class_ is, via it's _identifier_. It is messy though.

> Doc comments can contain references to various code elements. A library member
> or class member, etc. can be referenced by wrapping its name in square brackets

Reviewer: We try to [avoid use of etc](https://developers.google.com/style/word-list#etc) if possible.

```suggestion
Doc comments can contain references to various code elements.
Elements, such as functions and classes, can be referenced by wrapping their name in square brackets
```

Author: Done

(`[...]`) in a doc comment (a comment starting with `///`). Examples:

```dart
/// Returns a [String].
String f() => 'Hello';

/// Wraps [object] with [Future.value].
Future<T> g<T>(T object) => Future.value(object);
```

These doc comments contain references to the `String` class, the `object`
parameter, and the `Future.value` constructor.

## Features of references

> There are several benefits to referring to a code element in a doc comment as a
> doc comment reference:

Reviewer: I think we don't refer to a code element "as a reference", but "with a reference".

Also can we avoid the double "doc comment"?

```suggestion
There are several benefits to referring to code elements with doc comment references:
```

Author: Love it, done.

> ### Editor features

Reviewer: Perhaps the end of this section or a tip aside can call out the [`comment_references`](https://dart.dev/lints/comment_references) lint as enabling verification of them?

```
:::tip
<Tip content goes here>
:::
```

Another option would be calling this out in the "What can be referenced" section.

Author: Done, love it.

Doc comment references enable several IDE features. An element's name can be
code-completed within square brackets. When an element is renamed via an IDE
command, it can rewrite uses of that element, including references in doc
comments. When an IDE lists all "references" to an element, it can include
references in doc comments. An IDE can also provide Go-to-definition support at

> the location of a doc comment reference.

Reviewer: Optionally consider making this a list, so readers can quickly see the features supported if they know them, and choose to learn more. For example:

```
Doc comment references enable several IDE features:

- **Code completion**
  An element's name can be code-completed within square brackets.
- **Rename refactoring**
  When an element is renamed via an IDE command, it can
  rewrite uses of that element, including references in doc comments.
```

Author: Great idea, done.

### API documentation

> In API documentation generated by `dart doc`, a doc comment reference is

Author: Oops, just saw this! Done

linked, if possible, to the documentation page of the element being referenced.
If the element does not have a documentation page (for example, a function
parameter, a type variable, or a private class), then no link is created.

> ## What can be referenced

Author: Great idea, I've started an attempt at this.

> ## What can be referenced

Author: SG. I'm sure there's a good solution in here. Maybe just experiment with different table layouts.

Most library members can be referenced in a doc comment, including classes,
constants, enums, named extensions, extension types, functions, mixins, and
type aliases. This includes all in-scope library members, either declared

> locally, or imported. Library members which are imported with an import prefix

Author: Done

> can be referenced with the prefix. Examples:

Author: Done

```dart
import 'dart:math' as math;

/// [List] is in scope.
/// So is [math.max].
int x = 7;
```

Most members of a class, an enum, an extension, an extension type, and a mixin
can also be referenced. A reference to a member that is not in scope must be

> qualified with its container's name. For example the `wait` static method on

Author: I like it, thanks!

the `Future` class can be referenced in a doc comment with `[Future.wait]`.
This is true for instance members as well; the `add` method and the `length`
property on the `List` class can be referenced with `[List.add]` and
`[List.length]`. In-scope members like these can even be referenced without the

> qualifying container name:

Author: Great suggestion. Done.

```dart
abstract class MyList<E> implements List<E> {
  /// Refer to [add] and [contains], which is declared on [Iterable].
  void myMethod() {}
}
```

Parameters of a function and parameters of a function type can be referenced in
a doc comment only when they are in scope. They can therefore only be
referenced within a doc comment on such a parameter's function or on a type

> alias for such a paramter's enclosing function type.

Author: Done

Type parameters can be referenced in a doc comment only when they are in scope.
Therefore a type parameter of a method, top-level function, or type alias can
only be referenced within a doc comment on that element, and a type parameter
of a class, enum, extension, extension type, and mixin can only be referenced
within a doc comment on that element or on one of its members.

> The doc comment for a type alias which aliases a class, enum, extension type,

Author: Done

> or mixin cannot reference any of the aliased type's members, if it has any, as

Author: Done
